### ISSUE TITLE
impossible to use CP/M

### ISSUE THREAD

**Issue by dddfffggg777** [simonowen/simcoupe#86]
the issue https://github.com/simonowen/simcoupe/issues/64 still not solved
![image](https://user-images.githubusercontent.com/114140927/205469337-49f82fcf-6d50-4fc6-8cd3-11d1a65f5b53.png)
![image](https://user-images.githubusercontent.com/114140927/205469338-3d25ca62-7a7e-4287-8b9b-a88c778c1835.png)
impossible to use CP/M

**Comment by simonowen** [comment]
I believe I've reproduced this and understand what is happening. I missed this previously as I always formatted a new disk rather than an re-using an existing SAM-format disk.

B: maps to SAM floppy drive 1, which probably still contains your Pro-Dos boot disk (was that intentional?). I suspect it's probably in MGT format, which is a fixed-format sector-based container (2*80*10*512=819200 bytes). In that case the target is 10 sectors/track and you're trying to format it as CP/M 9 sectors/track, which is incompatible with the MGT container.  SimCoupe tries to fail this command by signalling a FDC WRITE_FAULT, but Pro-Dos doesn't notice and continues trying to format the remaining tracks. At the end of the process the target disk is completely unchanged as none of the track formatting worked. The directory listing will find the original (mostly) zero-filled SAM disk when it tries to show a directory listing, showing question marks instead of the NUL control character. This is the same as if you booted Pro-Dos and issued `dir b:` immediately.

I suspect that returning a write fault might not be valid controller behaviour so SimCoupe can't use this as a failure mechanism. The only failure guaranteed to work is reporting that the disk is write-protected, which has the potential to confuse users but at least it shows it hasn't worked. It's very much an emulator-specific problem due to the limitations of the target disk image container.

Some work-arounds that you could use with the latest official SimCoupe release:
- convert your MGT disk to EDSK (using `SAMdisk image.mgt image.dsk`) to give a flexible-format container that supports both 10-sector and 9-sector formats, which then can be reformatted successfully.
- Insert a new blank disk that supports CP/M format (File -> New, then pick EDSK or CP/M format). Or File -> Floppy 2 -> New for drive 2, which will be C: in Pro-Dos.
- Insert any other 9-sector disk image or any EDSK image you're happy to reformat.

I'll make a fix to instead report WRITE_PROTECT for any incompatible disk formatting.

**Comment by dddfffggg777** [comment]
format.com ignoring "no disk" state and format air

**Comment by dddfffggg777** [comment]
there is no SAMdisk documentation

**Comment by dddfffggg777** [comment]
SAMdisk.exe copy Pro-DOS-v2.dsk zzz.scl
Error: SCL is not supported for output

SAMdisk.exe copy Pro-DOS-v2.dsk zzz.trd
Error: TRD writing not implemented

----------------------
useless

**Comment by simonowen** [comment]
> format.com ignoring "no disk" state and format air same with dir command

Pro-Dos doesn't check for a disk before sending the WRITE_TRACK command and relies on a FDC quirk that keeps the controller busy until a disk is inserted. SimCoupe doesn't current support that stuck state but I can add it. I'd not seen this before as most programs checking for a disk before sending FDC commands.

I can't see a problem with the `dir b:` command without a disk. It loops waiting and the command continues once a disk is inserted. What problem are you seeing?

**Comment by dddfffggg777** [comment]
ZIP file with CP/M files most used CP/M disk format in internet

**Comment by simonowen** [comment]
There are many different flavours of CP/M and I'd be surprised if many available disk images are compatible directly compatible with Pro-Dos, though I don't know for sure. Some have a header on the disk to describe the format parameters but many lack it and expect the OS to know.

SimCoupe and SAMdisk support for CP/M disks doesn't really go beyond 720K DOS disks used by Pro-Dos, which are seen as a normal 9-sector double-density format. The rest of how they're interpreted is up to Pro-Dos, but anything that works in Pro-Dos on a real SAM machine should work in SimCoupe too.

I have fixes for the two issues you found, which I'll make available after I've had a chance to run some more tests with them.

**Comment by dddfffggg777** [comment]
here is example of CP/M disk:
https://winworldpc.com/download/146cc2a0-51c2-a9e2-84a2-11c3a6e28094
with i cant use

**Comment by simonowen** [comment]
That's not a CP/M disk it's just a zip archive containing some CP/M command and data files. To use the files with Pro-Dos in SimCoupe you'll need to find a way to create a CP/M disk image from them. The disk must also be created with the correct CP/M parameters that are compatible with Pro-Dos. I don't know what the parameters are and I don't know of a utility that will convert files to a disk image. You might be better asking the CP/M community and maybe looking at the Pro-Dos manual in case it can help you.

Only once you have a compatible CP/M disk image is it worth returning to try it in SimCoupe. Boot Pro-Dos then insert your new disk image and try a directory listing, etc.  The disk image will probably be a raw 720K file (737280 bytes).  Even if it doesn't work it's unlikely I'll be able to advise you further as my CP/M knowledge is very limited.

All this is well beyond the scope of SimCoupe support and this GitHub issue.  CP/M isn't a native SAM Coupé feature, it's just a software application that runs on the machine to provide a CP/M environment. I can only help with emulation related issues and features, such as any difference in behaviour between running on real SAM hardware and SimCoupe.

**Comment by dddfffggg777** [comment]
here is CP/M disk creator:
https://www.mathematik.uni-ulm.de/users/ag/yaze-ag/cdm.1.html
but it use only *.yaze format

**Comment by simonowen** [comment]
There seems to be a much simpler way to transfer files for use with Pro-Dos. It includes a FATREAD command that can read files from FAT16/FAT32 drives. If you format a USB stick as FAT32 and copy your files onto it you can make that drive available to Pro-Dos in SimCoupe.

**Launch SimCoupe as Administrator** to ensure it has raw access to the USB device. Then under Tools -> Options -> Drive 2, change the drive type to Atom Lite.  Instead of selecting a disk image path select your USB device from the drop-down list:

![image](https://user-images.githubusercontent.com/200089/206904697-6c002b1b-2e22-4f9d-ad1d-06b37234b639.png)

Then boot Pro-Dos and run `FATREAD C:`  (or whatever your USB stick show up as).  You'll then have a mini shell where you can view the files on the disk and copy to a different drive:

![simc0190](https://user-images.githubusercontent.com/200089/206904824-f30d8225-c46b-44c1-a01b-e60db2603bf4.gif)

Press Ctrl-C (in SimCoupe use the Right-Ctrl key on the keyboard) to exit the shell and return to Pro-Dos.  Then do whatever you need with the files.  In this example I'm copying the files to the RAM drive to run. You could also copy it to a .cpm disk image file for use in a later session, etc.

**Comment by dddfffggg777** [comment]
https://github.com/Othernet-Project/dir2fat32
this ok?

**Comment by dddfffggg777** [comment]
how i can create a fat32 image file?

**Comment by dddfffggg777** [comment]
how i can use the 2 disks? https://velesoft.speccy.cz/sam_zx128.htm

**Comment by dddfffggg777** [comment]
how about COM port emulation?
https://www.worldofsam.org/system/files/2018-09/SAM%20Coupé%20Comms%20Interface%20Manual.pdf
connect it to COM port of windows
can be used for wifi internet with ESP8266

**Comment by simonowen** [comment]
Using a USB stick is the simplest solution to import files into Pro-Dos, as described above. Creating a FAT32 image file just makes the process more complicated and is beyond the scope of the emulator and this issue.

Please keep to the topic of the issue in these threads. If you require general support on how to use disk images in SimCoupe see the manual, which explains how to insert and boot disks.  If you require more general SAM Coupé information and support I recommend joining the [SAM Coupe Users Group](https://www.facebook.com/groups/20486797963/) on Facebook.

You can open a new issue to request emulator features, such as support for the SAM communication interface. Though due to limited development time not all will be added, especially if they're not well supported by software and useful to many SimCoupe users.

**Comment by dddfffggg777** [comment]
software not come cose not supported, not supported cose dont have software

**Comment by simonowen** [comment]
That's true, but it's difficult to change that for such an old system. Only a couple of SAM programs were ever written to use the serial ports on the comms interface, and one of them was only available as a pre-release demo (Termite). The real comms interface also needed a hardware mod to reduce serial port noise and make it more reliable. For comparison, the parallel port in the comms interface was more widely used so SimCoupe supports it for DACs and printers.

I did have some experimental serial support in a very old (~20 years ago) development version of SimCoupe. It was Windows-only and supported just enough to talk to a device (such as a modem) on COM1. I didn't do anything more than send AT commands to it as even then most BBS servers had long since closed down.

Few modern PCs have a serial port (or USB to serial adapter) **and** something to connect to the serial port. The number of users that would make use of the emulator feature could probably be counted on one hand, so unfortunately it's likely to remain low on my ToDo list. With unlimited time I'd support every feature but in reality I have to prioritise those that benefit more users.

**Comment by dddfffggg777** [comment]
check about ESP8266 - a wifi to COM converter
its used now for old computers

**Comment by dddfffggg777** [comment]
https://github.com/mecparts/zmp
useful for file transfer

**Comment by dddfffggg777** [comment]
www gopher client for zx spectrum + esp8266 on COM port
https://www.youtube.com/watch?v=3wtMhGbxkl8
